tests/python/gaia-ui-tests/gaiatest/gaia_test.py

```diff
@@ -863,10 +863,26 @@ def cleanup_data(self):
         self.device.file_manager.remove('/data/local/storage/permanent')
         self.device.file_manager.remove('/data/local/storage/persistent')
         self.device.file_manager.remove('/data/local/storage/default')
-        self.device.file_manager.remove('/data/local/webapps')
         # remove remembered networks
         self.device.file_manager.remove('/data/misc/wifi/wpa_supplicant.conf')
 
+        if self.device.is_android_build:
+            file_items = self.device.file_manager.list_items('/data/local/webapps/')
+            apps = json.loads(self.device.file_manager.pull_file('/data/local/webapps/webapps.json'))
+            # All default apps have the same installTime, newly added webapps have a different installTime
+            # This difference is used here to selectively delete those webapp directories
+            # Also delete all directories that are not mentioned in the webapps.json file
+            for dir in file_items:
+                delete_file = True
+                for app in apps:
+                    if (dir == str(apps[app]['id']) and apps[app]['installTime'] == apps['system.gaiamobile.org']['installTime']):
+                        print(str(apps[app]['id']))
+                        delete_file = False
+                        break
+                if delete_file and self.device.file_manager.dir_exists('/data/local/webapps/' + dir):
+                    self.device.file_manager.remove('/data/local/webapps/' + dir)
+                    print('deleting dir for: ' + dir + ':: ' + str(apps[app]['installTime']))
```

Comment:
Yes, that works. I was afraid this wouldn't catch webapps that wouldn't be listed in the webapps.json file, for some reason, but it seems that doesn't happen (I thought it did, at some point).